numpy/dtypes.pyi

```diff
@@ -1,43 +1,641 @@
+    def hasobject(self) -> L[True]: ...
+    @property
+    def isbuiltin(self) -> L[1]: ...
+    @property
+    def name(self) -> L["object"]: ...
+    @property
+    def str(self) -> L["|O"]: ...
+
+# Flexible:
+
+@final
+class BytesDType(
+    Generic[_ItemSize_co],
+    _TypeCodes[L["S"], L["S"], L[18]],
+    _NoOrder,
+    _NBit[L[1],_ItemSize_co],
+    _BaseDType[np.bytes_],
+):
+    def __new__(cls, size: _ItemSize_co, /) -> BytesDType[_ItemSize_co]: ...
+    @property
+    def flags(self) -> L[0]: ...
+    @property
+    def hasobject(self) -> L[False]: ...
+    @property
+    def isbuiltin(self) -> L[0]: ...
+    @property
+    def name(self) -> LiteralString: ...
+    @property
+    def str(self) -> LiteralString: ...
+
+@final
+class StrDType(
+    Generic[_ItemSize_co],
+    _TypeCodes[L["U"], L["U"], L[19]],
+    _NativeOrder,
+    _NBit[L[4],_ItemSize_co],
+    _BaseDType[np.str_],
+):
+    def __new__(cls, size: _ItemSize_co, /) -> StrDType[_ItemSize_co]: ...
+    @property
+    def flags(self) -> L[8]: ...
+    @property
+    def hasobject(self) -> L[False]: ...
+    @property
+    def isbuiltin(self) -> L[0]: ...
+    @property
+    def name(self) -> LiteralString: ...
+    @property
+    def str(self) -> LiteralString: ...
+
+@final
+class VoidDType(
+    Generic[_ItemSize_co],
+    _TypeCodes[L["V"], L["V"], L[20]],
+    _NoOrder,
+    _NBit[L[1],_ItemSize_co],
+    _BaseDType[np.void],
+):
+    # NOTE: `VoidDType(...)` raises a `TypeError` at the moment
+    def __new__(cls, length: _ItemSize_co, /) -> NoReturn: ...
+    @property
+    def flags(self) -> L[0]: ...
+    @property
+    def hasobject(self) -> L[False]: ...
+    @property
+    def isbuiltin(self) -> L[0]: ...
+    @property
+    def name(self) -> LiteralString: ...
+    @property
+    def str(self) -> LiteralString: ...
+
+# Other:
+
+_DateTimeUnit_co = TypeVar(
+    "_DateTimeUnit_co",
+    bound=L[
+        "Y", "M", "W", "D",
+        "h", "m", "s", "ms", "us", "ns", "ps", "fs", "as",
+    ],
+    covariant=True,
+)
+
+@final
+class DateTime64DType(
+    Generic[_DateTimeUnit_co],
+    _TypeCodes[L["M"], L["M"], L[21]],
+    _NativeOrder,
+    _NBit[L[8], L[8]],
+    _BaseDType[np.datetime64],
+):
+    # NOTE: `DateTime64DType(...)` raises a `TypeError` at the moment
+    # TODO: Once implemented, don't forget the`unit: L["μs"]` overload.
+    def __new__(cls, unit: _DateTimeUnit_co, /) -> NoReturn: ...
+    @property
+    def flags(self) -> L[0]: ...
+    @property
+    def hasobject(self) -> L[False]: ...
+    @property
+    def isbuiltin(self) -> L[0]: ...
+    @property
+    def name(self) -> L[
+        "datetime64",
+        "datetime64[Y]",
+        "datetime64[M]",
+        "datetime64[W]",
+        "datetime64[D]",
+        "datetime64[h]",
+        "datetime64[m]",
+        "datetime64[s]",
+        "datetime64[ms]",
+        "datetime64[us]",
+        "datetime64[ns]",
+        "datetime64[ps]",
+        "datetime64[fs]",
+        "datetime64[as]",
```

Comment:
> Can't the unit be part of the class? Just like the byteorder.

Yea that's probably better. It's not really doing anything useful at the moment, and is mostly there for aesthetics.

> Also, for the byte-order shouldn't that refer to the order in the class definition?

Perhaps it should, but the only way I can think of to make that work, would be to add a byteorder type parameter, and add 28+ overloads in `__new__` (at least 2 for each unit).
But that might be a bit too messy, compared to the advantages.
... or isn't that what you mean?

Reply:
That was what I meant...  Anyway, since this is just about the DType classes as exposed here, let's not worry about it too much.  All of this might become more relvant when the classes are properly merged...